includes/utils/class-amp-dom-utils.php

```diff
@@ -711,4 +719,174 @@ public static function recursive_force_closing_tags( $dom, $node = null ) {
 	private static function is_self_closing_tag( $tag ) {
 		return in_array( strtolower( $tag ), self::$self_closing_tags, true );
 	}
+
+	/**
+	 * Check whether a given element has a specific class.
+	 *
+	 * @since 1.4.0
+	 *
+	 * @param DOMElement $element Element to check the classes of.
+	 * @param string     $class   Class to check for.
+	 * @return bool Whether the element has the requested class.
+	 */
+	public static function has_class( DOMElement $element, $class ) {
+		if ( ! $element->hasAttribute( 'class' ) ) {
+			return false;
+		}
+
+		$classes = $element->getAttribute( 'class' );
+
+		return (bool) preg_match( "/\b{$class}\b/", $classes );
+	}
+
+	/**
+	 * Get the ID for an element.
+	 *
+	 * If the element does not have an ID, create one first.
+	 *
+	 * @since 1.4.0
+	 *
+	 * @param DOMElement $element Element to get the ID for.
+	 * @param string     $prefix  Optional. Defaults to '_amp_wp_id_'.
+	 * @return string ID to use.
+	 */
+	public static function get_element_id( $element, $prefix = 'amp-wp-id' ) {
+		static $index_counter = [];
+
+		if ( $element->hasAttribute( 'id' ) ) {
+			return $element->getAttribute( 'id' );
+		}
+
+		if ( ! array_key_exists( $prefix, $index_counter ) ) {
+			$index_counter[ $prefix ] = 2;
+			$element->setAttribute( 'id', $prefix );
+
+			return $prefix;
+		}
+
+		$id = "{$prefix}-{$index_counter[ $prefix ]}";
+		$index_counter[ $prefix ] ++;
+
+		$element->setAttribute( 'id', $id );
+
+		return $id;
+	}
+
+	/**
+	 * Provides a "best guess" as to what XPath would mirror a given CSS
+	 * selector.
+	 *
+	 * This is a very simplistic conversion and will only work for very basic
+	 * CSS selectors. Therefore, it provides a filter for themes & plugins to
+	 * hook into to provide custom replacements.
+	 *
+	 * @since 1.4.0
+	 *
+	 * @param string $css_selector CSS selector to convert.
+	 * @return string|null XPath that closely mirrors the provided CSS selector,
+	 *                     or null if an error occurred.
+	 */
+	public static function xpath_from_css_selector( $css_selector ) {
```

Comment:
Given that this is very simplistic, and only used in the core theme sanitizer, how about moving it back into the core theme sanitizer and keep it `private` for now?

Reply:
Or else, rename it in a way that indicates it only supports basic selectors.

Reply:
Fixed via https://github.com/ampproject/amp-wp/pull/3403/commits/edafd6f6678c98a6cc213bd20daf169a5dcf15b9